### PR TITLE
feat(geoai): engine + CLI + docs + paper polish for v0.4.0

### LINKED ISSUE #46: Document reproducibility guarantees and their limits
**Context:** TerraFlow makes strong reproducibility claims (deterministic fingerprinting, seeded sampling, atomic writes). These are tested but not documented in one place.

**Problem:** A researcher adopting TerraFlow for a publication needs to know exactly what 'reproducible' means — what is guaranteed, and what is not (e.g., floating-point differences across OS/Python versions, pykrige solver non-determinism).

**Proposed Solution:** Add a `docs/reproducibility.md` page covering: what the fingerprint includes, what it excludes, known sources of non-determinism (pykrige variogram fit, scipy triangulation tie-breaking), and how to cite a run.

**Acceptance Criteria:**
- Page linked from README and MkDocs nav
- Known limitations stated honestly
- At least one known non-determinism source (pykrige variogram fit) documented explicitly

### LINKED ISSUE #94: docs(paper): cite Wu (2026) GeoAI JOSS paper
Citation slice of the `terraflow geoai` integration (parent epic #90).

Add the canonical Wu (2026) BibTeX entry to `paper/biblio.bib`:

```bibtex
@article{wu2026geoai,
  author  = {Wu, Qiusheng},
  year    = {2026},
  title   = {GeoAI: A Python package for integrating artificial intelligence
             with geospatial data analysis and visualization},
  journal = {Journal of Open Source Software},
  volume  = {11}, number = {118}, pages = {9605},
  doi     = {10.21105/joss.09605}
}
```

If the FTW pretrained model is shipped as the default for `engine: fields`, also add a `ftw_dataset` BibTeX entry.

Cite `@wu2026geoai` from a new "State of the Field" paragraph in `paper/paper.md` describing the `terraflow geoai` engine path; add a sentence to "Acknowledgements".

**Acceptance:** `manuscript.yml` workflow builds the JOSS PDF cleanly with the new citations resolved.

Refs #90

### LINKED ISSUE #96: chore(geoai): append commented geoai block to demo config and add changelog entry
Demo config + changelog slice of the `terraflow geoai` integration (parent epic #90). Depends on #93.

1. Append a commented-out `geoai:` block to `examples/demo_config.yml` so users can see the schema without enabling the heavy extra by default.
2. Add an entry under `## [Unreleased]` in `CHANGELOG.md` summarising the new subcommand, the optional `[geoai]` extra, the `GeoAIConfig` Pydantic block, the new fingerprint helper, the citation, and the new docs/notebook.

**Acceptance:** `terraflow geoai landcover -c examples/demo_config.yml` works after uncommenting the block (manual smoke test).

Refs #90

### LINKED ISSUE #97: ci(geoai): add opt-in geoai integration workflow gated on geoai_engine.py changes
CI slice of the `terraflow geoai` integration (parent epic #90).

Add `.github/workflows/geoai-tests.yml` that runs `pytest -m geoai_integration` against a real `[geoai]` install. Triggered only when files under `terraflow/geoai_engine.py` or `tests/test_geoai_engine.py` change (`paths:` filter), so the default CI matrix stays fast and PyTorch is downloaded only on PRs that actually touch the engine module.

Add the `geoai_integration` marker to `pyproject.toml` `[tool.pytest.ini_options]` so it does not warn.

**Acceptance:** workflow renders in `Actions` tab; manual `workflow_dispatch` succeeds on a tiny synthetic raster; default CI matrix is unchanged.

Refs #90

### PR DESCRIPTION
**Pull request by gmarupilla**
## Summary

Consolidated GeoAI integration sweep that lands the engine (#92), CLI subcommands (#93), all PR-checklist docs, opt-in CI workflow (#97), reproducibility doc additions (#46), and paper polish (#94) in a single PR.

Replaces PR #108 (closed; was opened under the wrong account). Supersedes PRs #106 and #107.

### Code
- **`terraflow/geoai_engine.py`** — `run_fields`, `run_landcover`, `run_canopy` orchestrators sharing one private `_run()` helper. Already-DRY in the original commit; no duplication concerns here.
- **`terraflow/cli.py`** — refactored: extracted `_config_option()` Annotated alias and `_invoke()` error helper, so every subcommand (run / sensitivity / validate / export / `geoai *`) shares one error ladder. **Drops SonarCloud duplication on the CLI sub-app from 12.5% to below the 3% threshold.**
- **`terraflow/__init__.py`** — re-export `compute_geoai_fingerprint` and `compute_run_fingerprint`.

### CI fixes
- Removed dynamic `_stub.calls` attribute pattern in `tests/test_geoai_engine.py` — factory now returns `(stub, calls)` tuple. Clears the three `Unused "type: ignore"` mypy errors on Py 3.11.
- Replaced `/tmp/x` test placeholder with `"extra"` in `tests/test_run_identity.py` — clears SonarCloud `python:S5443` critical vulnerability behind the D security rating.
- Strip ANSI codes before the geoai `--help` substring match in `tests/test_cli.py` — typer/click on Py 3.11 renders option names with embedded ANSI codes that broke the literal substring check.

### Documentation (PR-checklist closure)
- `docs/architecture/adr-007-geoai-engine.md` — design rationale, alternatives, consequences.
- `docs/geoai.md` — user guide.
- `docs/api/geoai_engine.md` — autodoc reference.
- `docs/notebooks/06_geoai_engine.ipynb` — end-to-end demo that stubs the ML deps so it executes without `[geoai]` installed.
- `docs/config/examples.md` — annotated `geoai:` block (closes #96).
- `docs/reproducibility.md` — new GeoAI fingerprint section documenting device + torch-minor sensitivity and CUDA non-determinism limits (closes #46).
- `mkdocs.yml` — nav entries under Architecture, Guides, API Reference, Notebooks.
- `README.md` — `[geoai]` extra in install table; GeoAI runners in At-a-Glance + CLI subcommands tables.
- `CHANGELOG.md` — `[Unreleased]` entries for engine, CLI, docs, notebook, paper, CI workflow.

### CI
- `.github/workflows/geoai-integration.yml` — opt-in workflow gated on paths under `terraflow/geoai_engine.py`, `tests/test_geoai_engine.py`, `terraflow/core/run_identity.py`. Installs `[geoai,dev]` extras and runs the GeoAI + fingerprint tests on Py 3.12 (closes #97).

### Paper polish
- Bumped submission date to 2026-06-08 and version reference to v0.4.0 in `paper/paper.md`.
- Added GeoAI paragraph citing the new `wu2026geoai` bib entry (closes #94).
- Removed the AI usage disclosure section per author preference.

### Repo hygiene
- `MASTER_PROMPT_*.md` + `*.local.md` patterns added to `.gitignore`.
- Removed accidental empty `docs/notebooks/Untitled.ipynb` + `Untitled1.ipynb`.
- Committed `uv.lock` (not previously in `.gitignore`).

## Test plan

- [x] `make lint` — ruff + black clean
- [x] `make typecheck` — mypy 0 errors on 18 source files
- [x] `make test-cov` — 283 pass / 1 skip, **87.46% coverage** (above 85% floor)
- [ ] CI: build-test on Py 3.10 / 3.11 / 3.12
- [ ] CI: SonarCloud Quality Gate (was D security + 12.5% duplication on PR #106/#107)
- [ ] CI: License check, Quality, Claude review, Docs preview, Docker e2e
- [ ] CI: opt-in `geoai-integration` workflow installs `[geoai,dev]` and passes

## Issues closed

Closes #46, #92, #93, #94, #95, #96, #97 and epic #90.

Supersedes #106 and #107.

## Out of scope

- **#67** (Docker smoke test passes without network access) — separate verification work; tracked as v0.4.1 follow-up.
- Real geoai-py inference bodies — engine `_do_*` functions remain `NotImplementedError` placeholders pending `geoai-py >= 0.2`; the orchestrator contract is fully exercised by the test suite via monkey-patching.

### CODE REVIEW

**Comment by gmarupilla** [comment]
## Review punch list applied (commit 03bd04b)

Address the multi-agent review:

**Blocking (4/4)**
1. ✅ `_invoke()` now translates `NotImplementedError` to exit 2 + "not yet implemented" message — fixes the stack-trace UX for users who run `terraflow geoai *` with `[geoai]` installed before #94 lands.
2. ✅ `geoai-integration.yml` `paths:` filter widened to include `terraflow/config.py`, `terraflow/cli.py`, and `tests/test_run_identity.py` so schema/CLI changes trigger the integration suite.
3. ✅ `geoai_manifest.json` now records `roi_applied: false` + `roi_note` explicitly — manifest is no longer silently misleading. ADR-007 + `docs/reproducibility.md` reference the flag.
4. ✅ Added direct unit tests for `_invoke()` covering ValueError / FileNotFoundError / ImportError / NotImplementedError / generic Exception + success pass-through.

**Doc nits (2/2)**
- ✅ `docs/geoai.md` artifact table aligned with the parametrized test set (`landcover.tif` + `landcover_proba.tif` + `class_fractions.json`; `canopy_height.tif` + `canopy_stats.json`).
- ✅ `docs/reproducibility.md` clarifies that `device` and `torch_major_minor` are *optional* in `compute_geoai_fingerprint` — engine always populates them; external callers must pin both for cross-tool agreement.

**Non-blocking deferred:**
- Notebook stub lambda → `def _stub` rewrite — current form executes cleanly under mkdocs-jupyter; cosmetic.

Local gates: ruff/black/mypy clean, 289 pass / 1 skip, 87.48% coverage.

File: terraflow/geoai_engine.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    ]
+    model_meta = _model_metadata(engine)
+
+    fingerprint = compute_geoai_fingerprint(config_dict, input_fingerprints, model_meta)
+    run_dir = Path(cfg.output_dir) / "runs" / fingerprint / "geoai"
+    manifest_path = run_dir / "geoai_manifest.json"
+
+    if manifest_path.exists():
+        logger.info(f"geoai cache hit — {run_dir}")
+        return run_dir
```
[the remark this answers is not in the document]
Comment (gmarupilla) [reply terraflow/geoai_engine.py:156]: Addressed in commit 1a2d567 — write order now puts `geoai_manifest.json` last so its existence is a true completion marker; engine artifacts and `report.json` are both written before it. Added `test_manifest_is_written_last_as_completion_marker` to pin the invariant. Local: 290 tests / 1 skipped.
